### PR TITLE
Fixed a memory issue when running the tunnel helpers

### LINKED ISSUE #2000: Memory leak on the tunnel helper
Current `devel` tunnel helper is leaking memory.

Apparently it only happens if exit mode is enabled, (either that or the lower amount of users on that kind of helper is not making the leak so obvious).

UPDATE: the non-exit tunnel helpers are also leaking, just at about half the rate than the others.

### LINKED ISSUE #2083: Session is not closing when running the tunnel helpers
When running the tunnel helpers (`python Tribler/community/tunnel/main.py`), the session is not closed when I quit the application by typing `q`. It keeps spamming and ultimately crashes:

```
INFO    1460452115.55            main:185   ONEXIT Waiting for Session to shutdown, will wait for 18 more seconds
INFO    1460452116.55            main:185   ONEXIT Waiting for Session to shutdown, will wait for 17 more seconds
INFO    1460452117.55            main:185   ONEXIT Waiting for Session to shutdown, will wait for 16 more seconds
INFO    1460452118.55            main:185   ONEXIT Waiting for Session to shutdown, will wait for 15 more seconds
INFO    1460452119.55            main:185   ONEXIT Waiting for Session to shutdown, will wait for 14 more seconds
INFO    1460452120.56            main:185   ONEXIT Waiting for Session to shutdown, will wait for 13 more seconds
INFO    1460452121.56            main:185   ONEXIT Waiting for Session to shutdown, will wait for 12 more seconds
INFO    1460452122.56            main:185   ONEXIT Waiting for Session to shutdown, will wait for 11 more seconds
INFO    1460452123.56            main:185   ONEXIT Waiting for Session to shutdown, will wait for 10 more seconds
```

### PR DESCRIPTION
**Pull request by devos50**
After a day of memory investigation, I found the root cause of the memory usage problem in the tunnel helpers when they are running for a long time.

The problem is that the Twisted log observer is intercepting lower-level Twisted events. When Twisted starts, a `LimitedHistoryLogObserver` is created, however, the name is very misleading since the observed history is not limited at all! An entry is written to the log, every time something interesting happens in the Tunnel community (for instance, the reactor starts or stops listening on a specific port or the `TunnelExitSocket` protocol starts).

Searching on the internet, brought me to this issue: [here](http://twistedmatrix.com/trac/ticket/7841). A workaround has been implemented [here](https://github.com/plq/neurons/blob/417836a729706f2d1449109ccbb7d4a510227050/neurons/daemon/config.py#L623). I basically took their approach with some minor changes. For instance, I don't replay the logs in `LimitedHistoryLogObserver`.

The fact that the memory is increasing faster when the community is running as exit node can easily be explained since we are opening more tunnels thus writing more log entries.

If you’re interested in how I found this exactly, I’ve written a blog post about it [here](http://blog.code-up.nl/finding-python-memory-peaks-using-meliae/).

I also fixed another issue in the tunnel helper main script that prevented the session from closing, fixing #2083.

Fixes #2000, fixes #2083

### CODE REVIEW

**Comment by devos50** [comment]
@whirm ready for review

**Comment by whirm** [comment]
I'd say this should be done as part of the startup process for any process (not just the tunnel helper) The tracker service will probably be affected too.

**Comment by whirm** [comment]
I would move it to dispersy so the Tracker twistd plugin can use it too and call it from all the twistd plugins in both dispersy and Tribler and from the logger setup function in Tribler.

**Comment by devos50** [comment]
@whirm that's ok, I can move it to Dispersy. Can you point out what's the best place to have this code in Dispersy?

**Comment by whirm** [comment]
I guess in `tool/`... dispersy itself won't be using it directly, only the dispersy tracker twistd plugin.

**Comment by devos50** [comment]
So we have this code in two places if I understand it correctly? In the tunnel helpers and the tracker twistd plugin in Dispersy?

**Comment by whirm** [comment]
No, the twistd plugin will be importing and calling the function, and as Tribler depends on dispersy, anything that needs to do the same on the Tribler repo can import and call the function too.

**Comment by devos50** [comment]
@whirm shall I open a new PR for calling the relevant code in dispersy?

**Comment by whirm** [comment]
nah, do it in this one

**Comment by whirm** [comment]
hm, you forgot to move the test to dispersy

**Comment by devos50** [comment]
@whirm see https://github.com/Tribler/dispersy/pull/466

**Comment by devos50** [comment]
@whirm I've updated this PR with the relevant changes.

File: Tribler/community/tunnel/main.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 from twisted.internet.task import LoopingCall
 from twisted.internet.stdio import StandardIO
+from twisted.logger import globalLogPublisher
```
Comment (whirm) [inline Tribler/community/tunnel/main.py]: This should be gone

Comment (devos50) [inline Tribler/community/tunnel/main.py]: @whirm `globalLogPublisher` is passed as root log object to the `clean_twisted_observers` method.

File: Tribler/dispersy [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit d056b388836cec227f6752683045534325b5b5e4
+Subproject commit 67c22ed23f0e079d775cbe6df05d907b1282fc24
```
Comment (whirm) [inline Tribler/dispersy]: I merged the test, so that should be 57ef0d210d98f2d9e41fca01fb48ef355efacf3a

File: Tribler/community/tunnel/main.py [hunk 3]
```diff
@@ -85,6 +88,8 @@ def clean_messages(self):
             if now - 3600 > self.crawl_message[k]['time']:
                 self.crawl_message.pop(k)
 
+        clean_twisted_observers(globalLogPublisher)
```
Comment (whirm) [inline Tribler/community/tunnel/main.py]: Oh, I just noticed that you keep passing the type to the function, it's a bit weird, but let's not send yet another PR to dispersy for that.

**Comment by devos50** [comment]
@whirm I will write a test tomorrow that starts and closes the tunnel helpers to see whether they are working. This also makes sure that the uncovered code is covered.

**Comment by whirm** [comment]
Nice!

**Comment by synctext** [comment]
Twisted has bad behavior or bug if it keeps 65k lines in memory. Even if you overwrite it.

Is an upstream fix possible? Lots of time?

**Comment by devos50** [comment]
@whirm ready for another review. Also, should I really cover that single line in the `LineReceiver` protocol?

**Comment by whirm** [comment]
@devos50 it's fine. I'll review it ASAP
